File: packages/core-ts/src/manifest.ts
import type { AgentManifest } from './types.js';

const CURRENT_MANIFEST_VERSION = '0.1';

const REQUIRED_FIELDS: (keyof AgentManifest)[] = [
  'manifest_version',
  'agent_id',
  'name',
  'version',
  'capabilities',
  'expires_at',
];

export interface CreateManifestConfig {
  agentId: string;
  name: string;
  version: string;
  capabilities: AgentManifest['capabilities'];
  expiresAt: string;
  operator?: AgentManifest['operator'];
  policyUrl?: string;
  metadata?: Record<string, unknown>;
}

/** Create a new agent manifest from a config object. */
export function createManifest(config: CreateManifestConfig): AgentManifest {
  const manifest: AgentManifest = {
    manifest_version: CURRENT_MANIFEST_VERSION,
    agent_id: config.agentId,
    name: config.name,
    version: config.version,
    capabilities: config.capabilities,
    expires_at: config.expiresAt,
  };

  if (config.operator) manifest.operator = config.operator;
  if (config.policyUrl) manifest.policy_url = config.policyUrl;
  if (config.metadata) manifest.metadata = config.metadata;

  const errors = validateManifest(manifest);
  if (errors.length > 0) {
    throw new Error(`Invalid manifest: ${errors.join('; ')}`);
  }

  return manifest;
}

/** Validate a manifest against the schema. Returns an array of error strings (empty if valid). */
export function validateManifest(manifest: AgentManifest): string[] {
  const errors: string[] = [];

  for (const field of REQUIRED_FIELDS) {
    if (manifest[field] === undefined || manifest[field] === null) {
      errors.push(`Missing required field: ${field}`);
    }
  }

  if (manifest.manifest_version && manifest.manifest_version !== CURRENT_MANIFEST_VERSION) {
    errors.push(
      `Unsupported manifest version: ${manifest.manifest_version} (expected ${CURRENT_MANIFEST_VERSION})`
    );
  }

  if (manifest.agent_id && !manifest.agent_id.startsWith('aid_ed25519_')) {
    errors.push('agent_id must start with "aid_ed25519_"');
  }

  if (manifest.name && manifest.name.length > 256) {
    errors.push('name must be 256 characters or less');
  }

  if (manifest.capabilities) {
    if (!Array.isArray(manifest.capabilities)) {
      errors.push('capabilities must be an array');
    } else {
      for (let i = 0; i < manifest.capabilities.length; i++) {
        const cap = manifest.capabilities[i];
        if (!cap.id) errors.push(`capabilities[${i}]: missing id`);
        if (!cap.description) errors.push(`capabilities[${i}]: missing description`);
      }
    }
  }

  if (manifest.expires_at) {
    const expiry = new Date(manifest.expires_at);
    if (isNaN(expiry.getTime())) {
      errors.push('expires_at is not a valid ISO 8601 date');
    }
  }

  return errors;
}

/**
 * Deterministic JSON serialization following RFC 8785 (JCS) conventions.
 * Produces byte-for-byte identical output for logically equivalent manifests
 * across TypeScript and Python SDKs.
 *
 * Cross-language contract:
 * - Object keys are sorted lexicographically
 * - Properties with `undefined` or `null` values are OMITTED (not serialized)
 * - Strings use JSON.stringify escaping
 * - Numbers follow ES2024 serialization rules
 * - Arrays preserve order; null/undefined elements become "null"
 */
export function canonicalizeManifest(manifest: AgentManifest): string {
  const cleaned = stripNullUndefined(manifest);
  return canonicalize(cleaned);
}

/**
 * Recursively remove all keys whose values are null or undefined,
 * ensuring consistent behavior with the Python SDK which filters None.
 */
function stripNullUndefined(value: unknown): unknown {
  if (value === null || value === undefined) return undefined;
  if (typeof value !== 'object') return value;
  if (Array.isArray(value)) {
    return value.map(stripNullUndefined);
  }
  const result: Record<string, unknown> = {};
  for (const [k, v] of Object.entries(value as Record<string, unknown>)) {
    if (v !== null && v !== undefined) {
      result[k] = stripNullUndefined(v);
    }
  }
  return result;
}

function canonicalize(value: unknown): string {
  if (value === undefined) return 'null';
  if (value === null) return 'null';
  if (typeof value === 'boolean') return value.toString();
  if (typeof value === 'number') return serializeNumber(value);
  if (typeof value === 'string') return JSON.stringify(value);
  if (Array.isArray(value)) {
    return '[' + value.map(canonicalize).join(',') + ']';
  }
  if (typeof value === 'object') {
    const keys = Object.keys(value as Record<string, unknown>).sort();
    const pairs = keys
      .filter((k) => (value as Record<string, unknown>)[k] !== undefined)
      .map((k) => JSON.stringify(k) + ':' + canonicalize((value as Record<string, unknown>)[k]));
    return '{' + pairs.join(',') + '}';
  }
  throw new Error(`Cannot canonicalize value of type ${typeof value}`);
}

function serializeNumber(n: number): string {
  if (Object.is(n, -0)) return '0';
  if (!isFinite(n)) throw new Error('Cannot canonicalize non-finite number');
  return JSON.stringify(n);
}

File: packages/core-py/agentid/manifest.py
from __future__ import annotations

import json
import math
from dataclasses import asdict
from datetime import datetime
from typing import Any

from agentid.types import AgentManifest, Capability, Constraint, Operator

MANIFEST_VERSION = "0.1"

REQUIRED_FIELDS = ["manifest_version", "agent_id", "name", "version", "capabilities", "expires_at"]


def create_manifest(
    *,
    agent_id: str,
    name: str,
    version: str,
    capabilities: list[Capability],
    expires_at: str,
    operator: Operator | None = None,
    policy_url: str | None = None,
    metadata: dict[str, Any] | None = None,
) -> AgentManifest:
    manifest = AgentManifest(
        manifest_version=MANIFEST_VERSION,
        agent_id=agent_id,
        name=name,
        version=version,
        capabilities=capabilities,
        expires_at=expires_at,
        operator=operator,
        policy_url=policy_url,
        metadata=metadata,
    )

    errors = validate_manifest(manifest)
    if errors:
        raise ValueError(f"Invalid manifest: {'; '.join(errors)}")

    return manifest


def validate_manifest(manifest: AgentManifest) -> list[str]:
    errors: list[str] = []
    d = asdict(manifest)

    for field_name in REQUIRED_FIELDS:
        val = d.get(field_name)
        if val is None or (isinstance(val, str) and val == ""):
            errors.append(f"Missing required field: {field_name}")

    if manifest.manifest_version and manifest.manifest_version != MANIFEST_VERSION:
        errors.append(
            f"Unsupported manifest version: {manifest.manifest_version} "
            f"(expected {MANIFEST_VERSION})"
        )

    if manifest.agent_id and not manifest.agent_id.startswith("aid_ed25519_"):
        errors.append('agent_id must start with "aid_ed25519_"')

    if manifest.name and len(manifest.name) > 256:
        errors.append("name must be 256 characters or less")

    if manifest.capabilities is not None:
        if not isinstance(manifest.capabilities, list):
            errors.append("capabilities must be a list")
        else:
            for i, cap in enumerate(manifest.capabilities):
                if not cap.id:
                    errors.append(f"capabilities[{i}]: missing id")
                if not cap.description:
                    errors.append(f"capabilities[{i}]: missing description")

    if manifest.expires_at:
        try:
            datetime.fromisoformat(manifest.expires_at.replace("Z", "+00:00"))
        except ValueError:
            errors.append("expires_at is not a valid ISO 8601 date")

    return errors


def canonicalize_manifest(manifest: AgentManifest) -> str:
    """RFC 8785 (JCS) canonicalization."""
    d = _manifest_to_dict(manifest)
    return _canonicalize(d)


def _manifest_to_dict(manifest: AgentManifest) -> dict[str, Any]:
    d = asdict(manifest)
    return _strip_none(d)


def _strip_none(value: Any) -> Any:
    if value is None:
        return None
    if isinstance(value, dict):
        stripped = {}
        for k, v in value.items():
            sv = _strip_none(v)
            if sv is not None and sv != {}:
                stripped[k] = sv
        return stripped
    if isinstance(value, list):
        return [_strip_none(item) for item in value]
    return value


def _canonicalize(value: Any) -> str:
    if value is None:
        return "null"
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, int) and not isinstance(value, bool):
        return str(value)
    if isinstance(value, float):
        if math.isnan(value) or math.isinf(value):
            raise ValueError("Cannot canonicalize non-finite number")
        if value == int(value):
            return str(int(value))
        return json.dumps(value)
    if isinstance(value, str):
        return json.dumps(value)
    if isinstance(value, list):
        return "[" + ",".join(_canonicalize(item) for item in value) + "]"
    if isinstance(value, dict):
        keys = sorted(value.keys())
        pairs = [
            json.dumps(k) + ":" + _canonicalize(value[k])
            for k in keys
            if value[k] is not None
        ]
        return "{" + ",".join(pairs) + "}"
    raise TypeError(f"Cannot canonicalize value of type {type(value)}")
